tools/querytee/splitting_handler.go

```diff
@@ -0,0 +1,211 @@
+	return queryrangebase.MergeMiddlewares(middleware...).Wrap(preferredRT)
+}
+
+// ServeHTTP implements http.Handler interface to serve queries that can be split.
+// If ComparisonMinAge is 0 (legacy mode), serve the non-splitting fan-out handler directly.
+// If ComparisonMinAge > 0 (splitting mode), it wraps the fan-out handler with engineRouter
+// middleware to serve split queries based on data age.
+func (f *SplittingHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	_, ctx, err := tenant.ExtractTenantIDFromHTTPRequest(r)
+	if err != nil {
+		level.Error(f.logger).Log(
+			"msg", "failed to extract tenant ID",
+			"err", err,
+			"req", r.URL.String(),
+		)
+		http.Error(w, err.Error(), http.StatusUnauthorized)
+		return
+	}
+
+	// The codec decode/encode cycle loses custom headers, so we preserve them for downstream
+	headersCopy := r.Header.Clone()
+	ctx = context.WithValue(ctx, originalHTTPHeadersKey, headersCopy)
+
+	req, err := f.codec.DecodeRequest(ctx, r, nil)
+	if err != nil {
+		query := r.Form.Get("query")
+		level.Warn(f.logger).Log(
+			"msg", "failed to decode request",
+			"query", query,
+			"req", r.URL.String(),
+			"err", err,
+		)
+		server.WriteError(err, w)
+		return
+	}
+
+	var resp queryrangebase.Response
+	if f.goldfishManager != nil && f.goldfishManager.ComparisonMinAge() > 0 {
+		resp, err = f.serveSplits(ctx, req)
+	} else {
+		resp, err = f.fanOutHandler.Do(ctx, req)
+	}
+
+	if err != nil {
+		switch r := resp.(type) {
+		case *NonDecodableResponse:
+			http.Error(w, string(r.Body), r.StatusCode)
+		default:
+			level.Warn(f.logger).Log("msg", "handler failed", "err", err)
+			server.WriteError(err, w)
+		}
+		return
+	}
+
+	// Encode the response back to HTTP
+	httpResp, err := f.codec.EncodeResponse(ctx, r, resp)
+	if err != nil {
+		level.Warn(f.logger).Log("msg", "failed to encode response", "err", err)
+		server.WriteError(err, w)
+		return
+	}
+	defer httpResp.Body.Close()
+
+	// Copy response headers
+	for key, values := range httpResp.Header {
+		for _, value := range values {
+			w.Header().Add(key, value)
+		}
+	}
+
+	// Write status code
+	w.WriteHeader(httpResp.StatusCode)
+
+	// Copy response body
+	if _, err := io.Copy(w, httpResp.Body); err != nil {
+		level.Warn(f.logger).Log("msg", "unable to write response body", "err", err)
+	}
+}
+
+func (f *SplittingHandler) serveSplits(ctx context.Context, req queryrangebase.Request) (queryrangebase.Response, error) {
+	switch op := req.(type) {
+	case *queryrange.LokiRequest:
+		if op.Plan == nil {
+			err := errors.New("query plan is empty")
+			query := req.GetQuery()
+			level.Warn(f.logger).Log("msg", "query plan is empty", "query", query, "err", err)
+			return nil, err
+		}
+
+		switch op.Plan.AST.(type) {
+		case syntax.VariantsExpr, syntax.SampleExpr:
+			return f.metricsQueryHandler.Do(ctx, req)
+		default:
+			return f.logsQueryHandler.Do(ctx, req)
+		}
+	default:
+		return f.logsQueryHandler.Do(ctx, req)
```

Comment:
nit: not a blocking comment as this behavior has not changed with the refactor.

for query types that are not supported, fallback to fanout without splitting or only sent to preferred backend?